### PR TITLE
Remove buildClassBorrowMethod

### LINKED ISSUE #13438: removing buildClassBorrowMethod
PR #10567 added buildClassBorrowMethod to enable .borrow() for unmanaged types. However this has the bad side-effect of using virtual dispatch to do what effectively amounts to an identity operation. (My concern with that is compilation performance and program performance).

Instead, the compiler should handle such borrow calls at resolution time (as it handles things like isSubtype). This will prohibit user-defined .borrow methods on classes, but supporting those was not intended in the original design #9501.

### PR DESCRIPTION
**Pull request by e-kayrakli**
Resolves #13438 

All credit goes to @mppf 

Test:
- [x] Full quickstart (see below for exception)

### CODE REVIEW

**Comment by e-kayrakli** [comment]
`test/classes/delete-free/borrowed/user-defined-borrow` fails. I think it is supposed to fail and should be removed with this PR.

However, should the word `borrow` be made a reserved keyword to prevent users for writing a `borrow` method?

**Comment by mppf** [comment]
Re test/classes/delete-free/borrowed/user-defined-borrow, we can solve it by making the compiler give an error if a borrow method is defined for a class. This can happen in scopeResolve.

**Comment by e-kayrakli** [comment]
I will take a look at scopeResolve and ask if I cannot figure out.  Also, I am aware of the Travis failure, it looks like it is trivial

**Comment by e-kayrakli** [comment]
Last batch of commits

- throw an error message when a user defined class has a method named `borrow`
- makes travis happy
- does NOT modify the failing test. It should either be removed or adjusted so that it tests the error message.

The branch still passes full quickstart paratest (minus the test in question)

@mppf -- Is the way that I am doing the check correct? Do you have a better error message suggestion? Any other feedback?

File: compiler/passes/scopeResolve.cpp [hunk 1]
```diff
@@ -2404,6 +2404,18 @@ static void removeUnusedModules() {
   }
 }
 
+static void detectUserDefinedBorrowMethods() {
+  forv_Vec(FnSymbol, fn, gFnSymbols) {
+    if (fn->isMethod()) {
+      if (strncmp(fn->name, "borrow", 6) == 0) {
```
Comment (cassella) [inline compiler/passes/scopeResolve.cpp]: Won't this flag any method whose name starts with `"borrow"`?

Comment (mppf) [reply compiler/passes/scopeResolve.cpp]: It's better to not use `strcmp` or `strncmp` in the compiler if you can avoid it. In this case, you can just do this:

``` c++
const char astrBorrow = astr("borrow");
forv_Vec(FnSymbol, fn, gFnSymbols) {
    if (fn->isMethod() && fn->name == astrBorrow) {
```

File: compiler/passes/scopeResolve.cpp [hunk 2]
```diff
@@ -2404,6 +2404,18 @@ static void removeUnusedModules() {
...
+        if (!fn->_this->type->symbol->hasFlag(FLAG_MANAGED_POINTER)) {
```
Comment (mppf) [inline compiler/passes/scopeResolve.cpp]: Please add `isClassLike(fn->_this->type) && !fn->_this->type->symbol->hasFlag(FLAG_MANAGED_POINTER)`.

**Comment by e-kayrakli** [comment]
Last commit addressed the feedback. The two tests that this PR is concerned with and primers passed with quickstart.

File: compiler/resolution/functionResolution.cpp [hunk 3]
```diff
@@ -2462,6 +2466,50 @@ static bool resolveBuiltinCastCall(CallExpr* call)
   return false;
 }
 
+static bool resolveClassBorrowMethod(CallExpr* call) {
+  if (isUnresolvedSymExpr(call->baseExpr)) {
+    if (call->numActuals() == 2) { //mt, this
+      if (call->isNamed("borrow") && dtMethodToken == call->get(1)->typeInfo()) {
```
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: Let's make this do pointer comparison on the UnresolvedSymExpr* -> unresolved. To do that, you'll add a `const char* astrBorrow;` in symbol.h / symbol.c and initialize it just like e.g. astrPostfixBang

Comment (e-kayrakli) [reply compiler/resolution/functionResolution.cpp]: I got the creating and comparing agains `astrBorrow` part. And I think you want me to convert the `call->baseExpr` to `UnresolvedSymExpr` and call `isNamed` on it?

Comment (mppf) [reply compiler/resolution/functionResolution.cpp]: You can call isNamedAstr or `myUnresolvedSymbol->unresolved == myAstr`

Comment (e-kayrakli) [reply compiler/resolution/functionResolution.cpp]: Ah, didn't know `unresolved` was a field. Will do

Comment (e-kayrakli) [reply compiler/resolution/functionResolution.cpp]: I liked isNamedAstr better.

File: compiler/passes/scopeResolve.cpp [hunk 4]
```diff
@@ -2404,6 +2404,19 @@ static void removeUnusedModules() {
   }
 }
 
+static void detectUserDefinedBorrowMethods() {
+  const char *astrBorrow = astr("borrow");
```
Comment (mppf) [inline compiler/passes/scopeResolve.cpp]: Maybe use the global astrBorrow here too?

Comment (e-kayrakli) [reply compiler/passes/scopeResolve.cpp]: oh, duh :(

**Comment by e-kayrakli** [comment]
@mppf -- is this ready to be merged?

**Pull request merged**